src/realm/object_converter.cpp

```diff
@@ -505,73 +514,83 @@ void InterRealmValueConverter::handle_dictionary_in_mixed(Dictionary& src_dictio
+    if (src_any != dst_any)
+        return false;
+    if (!is_collection(src_any))
+        return true;
+    // Check collections are equal
+    if (src_any.is_type(type_List)) {
+        auto src_element = src_list.get_list(ndx_src);
+        auto dst_element = dst_list.get_list(ndx_dst);
+        auto src_element_size = src_element->size();
+        auto dst_element_size = dst_element->size();
+        if (src_element_size != dst_element_size)
             return false;
-        for (size_t i = 0; i < size_src; ++i) {
-            auto src_mixed = nested_src_list->get_any(i);
-            auto dst_mixed = nested_dst_list->get_any(i);
-            if (src_mixed != dst_mixed)
+        for (size_t i = 0; i < src_element_size; ++i) {
+            if (!(check_matching_list(*src_element, *dst_element, i, i)))
                 return false;
         }
     }
-    else if (type == CollectionType::Dictionary) {
-        auto nested_src_dictionary = src_list.get_dictionary(ndx_src);
-        auto nested_dst_dictionary = dst_list.get_dictionary(ndx_dst);
-        auto size_src = nested_src_dictionary->size();
-        auto size_dst = nested_dst_dictionary->size();
-        if (size_src != size_dst)
+    else if (src_any.is_type(type_Dictionary)) {
+        auto src_element = src_list.get_dictionary(ndx_src);
+        auto dst_element = dst_list.get_dictionary(ndx_dst);
+        auto src_element_size = src_element->size();
+        auto dst_element_size = dst_element->size();
+        if (src_element_size != dst_element_size)
             return false;
-        for (size_t i = 0; i < size_src; ++i) {
-            auto [src_key, src_mixed] = nested_src_dictionary->get_pair(i);
-            auto [dst_key, dst_mixed] = nested_dst_dictionary->get_pair(i);
+        for (size_t i = 0; i < src_element_size; ++i) {
+            auto src_key = src_element->get_key(i);
+            auto dst_key = dst_element->get_key(i);
             if (src_key != dst_key)
                 return false;
-            if (src_mixed != dst_mixed)
+            if (!check_matching_dictionary(*src_element, *dst_element, src_key.get_string()))
                 return false;
         }
     }
     return true;
 }
 
 bool InterRealmValueConverter::check_matching_dictionary(const Dictionary& src_dictionary,
-                                                         const Dictionary& dst_dictionary, StringData key,
-                                                         CollectionType type) const
+                                                         const Dictionary& dst_dictionary, StringData key) const
 {
-    if (type == CollectionType::List) {
-        auto n_src_list = src_dictionary.get_list(key);
-        auto n_dst_list = dst_dictionary.get_list(key);
-        auto size_src = n_src_list->size();
-        auto size_dst = n_dst_list->size();
-        if (size_src != size_dst)
+    REALM_ASSERT(src_dictionary.contains(key) && dst_dictionary.contains(key));
```

Comment:
Should we just return `false` here instead of an assert?

Reply:
it's sort of a problem if the key (and the indexes in check_matching_list) is not in the dictionary. the caller ensures that (haven't really changed that). I added these to mostly catch unexpected behavior during development (i could turn it into debug asserts if it makes more sense).

Reply:
Agreed.